### PR TITLE
HPCC-23027 Return channels in ESP machine responses

### PR DESCRIPTION
**Pull request by wangkx**
For thor slave, the value is read from @channelsPerSlave.
For roxie, the value is read from @channelsPerNode.

A bug is fixed for retrieving thor slaves (processes) in
ws_machineEx::getThorProcesses(). The existing code uses
thor group name to lookup thor node groups. Then, one
slave process is defined for each node group. That is
incorrect if one thor slave has multiple channels because
each channel has its own node group. The new code calls
the getClusterProcessNodeGroup(). The call returns node
groups, one for each slave node. Then, the code defines
thor slaves in each node using @slavePerNode.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 1]
```diff
@@ -1570,6 +1576,9 @@ void CTpWrapper::getMachineList(const char* MachineType,
         if (!root)
             throw MakeStringExceptionDirect(ECLWATCH_CANNOT_GET_ENV_INFO, MSG_FAILED_GET_ENVIRONMENT_INFO);
 
+        bool hasPropChannelsPerNode = root->hasProp("@channelsPerNode");
```
Comment (afishbeck) [inline esp/smc/SMCLib/TpWrapper.cpp]: You could move these down to closer to where they are used, or do them inline.

**Review by afishbeck — COMMENTED** [review]
@wangkx some minor comments.

File: esp/smc/SMCLib/TpWrapper.cpp (hunk 1 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.cpp]: It is used inside a while loop. It may not be efficient if it is called many times.

Comment (afishbeck) [reply esp/smc/SMCLib/TpWrapper.cpp]: Ah, ok, that's fine.

**Comment by wangkx** [comment]
@afishbeck please review my changes.

**Review by afishbeck — APPROVED** [review]
@wangkx looks good.

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged.

**Pull request merged**